### PR TITLE
h265parse: enable the parser

### PR DESCRIPTION
**Pull request by dabrain34**
Can now play 2 frames with vk_video_samples

### CODE REVIEW

**Comment by ceyusa** [comment]
ouch... a couple lines here require gstreamer 1.21.0.1 (master)

not sure if this could be nice for users... I would rather stick with 1.20

**Comment by ceyusa** [comment]
perhaps we should add a simple ci

**Comment by dabrain34** [comment]
> perhaps we should add a simple ci

Any idea of a project using a runner which owns a gstreamer install or at least able to generate your docker image ?

**Comment by dabrain34** [comment]
> ouch... a couple lines here require gstreamer 1.21.0.1 (master)
Indeed but the H265 parser needs https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/2575
> 
> not sure if this could be nice for users... I would rather stick with 1.20
Need to stick on 1.21 or the next 1.22 release

**Comment by dabrain34** [comment]
This branch can now build > 1.20 including the change https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/3072

File: lib/vkvideo_parser/vulkan_interfaces.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+#ifdef VK_API_USE_DRIVER_REPO
+// If using the local driver repo with Vulkan APIs
+#include "vulkan/vulkannv.h"
+#else
+// Using the Vulkan APIs from Vulkan SDK
+#ifndef VK_ENABLE_BETA_EXTENSIONS
+#define VK_ENABLE_BETA_EXTENSIONS 1
+#endif
+#include "vulkan/vulkan.h"
+#endif
```
Comment (ceyusa) [inline lib/vkvideo_parser/vulkan_interfaces.h:26]: Why is this needed? I can't see any usage in our case?

File: test/dump.cpp [hunk 2]
```diff
@@ -869,11 +870,23 @@ dump_parser_sequence_info (const VkParserSequenceInfo * info)
...
+{
+  if (entry) {
+    start_object ("VkPicIf");
+    print_integer ("decodeWidth", entry->decodeWidth);  // frame_num(short-term) or LongTermFrameIdx(long-term)
+    print_integer ("decodeHeight", entry->decodeHeight);  // 0=short term reference, 1=long term reference
+    print_integer ("decodeSuperResWidth", entry->decodeSuperResWidth);  // non-existing reference frame (corresponding PicIdx
+    end_object ();
+  }
+}
+
```
Comment (ceyusa) [inline test/dump.cpp:884]: Just for the sake of "atomicity and clearness"  I would move these changes in another commit.

File: lib/h264dec.cpp [hunk 3]
```diff
@@ -19,7 +19,6 @@
 
 #include <atomic>
 
-#include "base.h"
```
Comment (ceyusa) [inline lib/h264dec.cpp:22]: Ouch... it would be possible to split this also? a commit related with importing headers, and other with h265 parsing?

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -25,3 +25,6 @@ jobs:
           TERM: dumb
         run:
           ninja -C ~/builddir
+      - name: Test H264
```
Comment (ceyusa) [inline .github/workflows/main.yml:28]: can we move these tests to another PR?

File: lib/h265dec.cpp [hunk 5]
```diff
@@ -561,17 +1020,24 @@ gst_h265_dec_update_picture_parameters (GstH265Decoder * decoder,
...
     case GST_H265_NAL_VPS:{
       GstH265VPS *vps = static_cast < GstH265VPS * >(nalu);
-      // if (pps_cmp (&self->last_pps, pps))
-      //   return;
-      // self->last_pps = *pps;
-      // fill_vps (pps, &self->vkp);
+      if (vps_cmp (&self->last_vps, vps))
+         return;
+      self->last_vps = *vps;
+      /* FIXME: The VPS arrives first but the vk video decoder expects the SPS to be provided first.
```
Comment (ceyusa) [inline lib/h265dec.cpp:1061]: Does this mean that currently no VPS nal is passed to decoder?

Comment (ceyusa) [reply lib/h265dec.cpp:1061]: it's ok if it's not, just to be aware of it.

File: lib/vkvideo_parser/vulkan_interfaces.h (hunk 1 above)
Comment (ceyusa) [reply lib/vkvideo_parser/vulkan_interfaces.h:26]: I understand now that it's required to avoid modifying lib/vkvideo_parser/VulkanVideoParserIf.h

it's ok.. but this header is horrible :)

File: lib/h265dec.cpp (hunk 5 above)
Comment (dabrain34) [reply lib/h265dec.cpp:1061]: yes the VPS is discarded by the NV parser, so I decided to remove it. That's why I put a FIXME

Comment (dabrain34) [reply lib/h265dec.cpp:1061]: If the VPS is processed then vk_video_samples is complaining that the SPS has not been received and assert

File: lib/vkvideo_parser/vulkan_interfaces.h (hunk 1 above)
Comment (dabrain34) [reply lib/vkvideo_parser/vulkan_interfaces.h:26]: yes I agree :)

**Comment by dabrain34** [comment]
tested with https://github.com/nvpro-samples/vk_video_samples/commit/b95d9f8b9a27cf421f7c037642d2c0c89d295f45
